src/main/java/htsjdk/samtools/CigarOperator.java

```diff
@@ -89,114 +117,116 @@
+     */
+    public static CigarOperator fromBamCode(final int bamCode) {
+        if (bamCode >= 0 && bamCode < bamCode2Value.length) {
+            return bamCode2Value[bamCode];
+        } else {
+            throw new IllegalArgumentException("Unrecognized CigarOperator: " + bamCode);
         }
     }
 
     /**
      *
      * @param e CigarOperator enum value.
      * @return CIGAR operator corresponding to the enum value in binary form as appears in a BAMRecord.
+     * @deprecated use {@link #bamCode()} instead.
      */
+    @Deprecated
     public static int enumToBinary(final CigarOperator e) {
-        switch(e) {
-            case M:
-                return OP_M;
-            case I:
-                return OP_I;
-            case D:
-                return OP_D;
-            case N:
-                return OP_N;
-            case S:
-                return OP_S;
-            case H:
-                return OP_H;
-            case P:
-                return OP_P;
-            case EQ:
-                return OP_EQ;
-            case X:
-                return OP_X;
-            default:
-                throw new IllegalArgumentException("Unrecognized CigarOperator: " + e);
-        }
+        return e.ordinal();
     }
 
-    /** Returns the character that should be used within a SAM file. */
+    /**
+     * Returns the binary representation of this operator as it appears in the BAMs.
+     * @return a unique value between 0 and 8.
+     */
+    public int bamCode() {
+        return ordinal();
+    }
+
+    /** Returns the character that should be used within a SAM file.
+     *
+     * @deprecated use {@link #asByte()} or {@link #asChar()}
+     */
+    @Deprecated
     public static byte enumToCharacter(final CigarOperator e) {
-        return e.character;
+        return (byte) e.character;
+    }
+
+    /**
+     * Returns the operator in its single byte representation.
+     * @return same as the first character (cast to byte) of the string returned by {@link #toString()}.
+     */
+    public byte asByte() {
+        return (byte) character;
+    }
+
+    /**
+     * Returns the operator in its single byte representation.
+     * @return same as the first character of the string returned by {@link #toString()}.
+     */
+    public char asChar() {
+        return character;
     }
 
     /** Returns true if the operator is a clipped (hard or soft) operator */
     public boolean isClipping() {
-        return this == S || this == H;
+        return (CLIP_MASK & mask) != 0;
     }
 
     /** Returns true if the operator is a Insertion or Deletion operator */
     public boolean isIndel() {
-        return this == I || this == D;
+        return (INDEL_MASK & mask) != 0;
     }
 
     /** Returns true if the operator is a Skipped Region Insertion or Deletion operator */
     public boolean isIndelOrSkippedRegion() {
-        return this == N || isIndel();
+        return (mask & INDEL_OR_SKIP_MASK) != 0;
```

Comment:
Do you ever see these show up in a profiler?  This looks like it should be a bit faster if the compiler can't write clever if statements.  I'm curious if this is a real performance problem though.

Reply:
If the code that use these enum it itself efficiently written, this probably won't show in a profiler. This mask based solution is an the Inlined version of a EnumSet... . 

I was curious myself so I did some bench marking. if you allow JIT to do its magic the mask and using an EnumSet is about the same cost say (1000 units) whereas compare with three constants (== || == || ==) was around 5000 2 constants 4500 (== || ==) and one only 2000.  
If you simply add a extra boolean field to identify members of a group then it is ~750. The same if your set happens to be begining or end of the enum const list so that you can use this.ordinal() <= X.ordinal() then it was around 750. It is easy to see how the compiler may transform the latter into the former and so the similar performance. 

Now disabling JIT (java -Djava.compiler=NONE)  all the solutions were the same cost except from using the EnumSet that was 4 or 5 times slower... so it seems that another factor was dominating for the faster alternatives (perhaps memory access). 

Anyway the bottom line is that using additional boolean fields is probably the faster way closely followed by using a mask and or EnumSet (but only if we can assume JIT is active).

Which one do you prefer?

Reply:
Interesting.  Let's just embed a bunch of constants then, that seems fast and readable.  We can set them in the constructor using the clear logic that exists now.

Reply:
Not sure what you mean... do you mean create additional ```final boolean``` fields (one per set) caching whether the value belongs to a particular set?

Reply:
In the new commit I done something different to make the constructor expression a bit more readable... is difficult to get what is going on with so many ""anonymous"" additional boolean arguments